### ISSUE TITLE
Im2col slow on big matrices

### ISSUE THREAD

**Issue by kali** [sonos/tract#570]

**Comment by kali** [comment]
An update here. We are working it, low priority, but we are on it.

Issue comes from the size of the im2col matrix: tract is typically optimized for small-ish models, while this is a model that has some *big* convolution. In this case, our strategy of splitting im2col+pack from the actual computation may be counterproductive. 

ONNX runtime handles thing differently  and gets better performance. First, they makes several pass on k over the output matrix. 

They also segment the im2col matrix in 128x128 blocks, only instantiating the im2col segment before packing them when they are required by the computation pass.

We will be experimenting with these strategies, how much they yield, and see if we can make them fit nicely in tract.

**Comment by kali** [comment]
Hello there, it's been a bit of a journey, but we have refactored how we were processing im2col for big convolutions. It is not merged on `main`. I am curious to know how you measure the impact on your model. Just out of the box first, but as a a bonus side effect, the new im2col implementation is now part of the product itself, so if you rebase the multithreading code on top of the current main, the im2col should happen in parallel too.

Hope you can share some result :)

**Comment by kali** [comment]
Last call to @ro99... are you still interested in this ?

**Comment by ro99** [comment]
Hello @kali I am working in something else right now and unfortunely I do not have much time left to work on this, please feel free to take over.

**Comment by kali** [comment]
OK, this issue was about non-multithreading optimisations to im2col (they should benefit multithread impl too) so I'm closing it now, I think we are were we can reasonably expect to be.
